dask/array/core.py

```diff
@@ -332,6 +362,139 @@ def broadcast_dimensions(argpairs, numblocks, sentinels=(1, (1,)),
     return valmap(first, g2)
 
 
+def _top(func, output, output_indices, *arrind_pairs, **kwargs):
+    """ Create a TOP symbolic mutable mapping, given the inputs to top
+
+    This is like the ``top`` function, but rather than construct a dict, it
+    returns a symbolic TOP object.
+
+    See Also
+    --------
+    top
+    TOP
+    """
+    numblocks = kwargs.pop('numblocks')
+    concatenate = kwargs.pop('concatenate', None)
+    new_axes = kwargs.pop('new_axes', {})
+
+    graph = ShareDict()
+
+    # Transform indices to canonical elements
+    # We use terms like _0, and _1 rather than provided index elements
+    arrind_pairs = list(arrind_pairs)
+    unique_indices = set(output_indices) | {i for ii in arrind_pairs[1::2] if ii is not None for i in ii}
+    sub = {k: atop_token(i, '.') for i, k in enumerate(sorted(unique_indices))}
+    output_indices = index_subs(tuple(output_indices), sub)
+    arrind_pairs[1::2] = [tuple(a) if a is not None else a for a in arrind_pairs[1::2]]
+    arrind_pairs[1::2] = [index_subs(a, sub) for a in arrind_pairs[1::2]]
+    new_axes = {index_subs(k, sub)[0]: v for k, v in new_axes.items()}
+
+    # Unpack dask values in non-array arguments
+    argpairs = list(partition(2, arrind_pairs))
+    for i, (arg, ind) in enumerate(argpairs):
+        if ind is None:
+            arg2, dsk2 = to_task_dask(arg)
+            if dsk2:
+                graph.update(dsk2)
+                argpairs[i] = (arg2, ind)
+
+    inputs = tuple([name for name, _ in argpairs])
+    inputs_indices = tuple([index for _, index in argpairs])
+
+    # Unpack delayed objects in kwargs
+    if kwargs:
+        kwargs, dsk_kwargs = to_task_dask(kwargs)
+
+        # replace keys in kwargs with _0 tokens
+        new_keys = list(core.get_dependencies(dsk_kwargs, task=kwargs))
+        new_tokens = tuple(atop_token(i) for i in range(len(inputs), len(inputs) + len(new_keys)))
+        sub = dict(zip(new_keys, new_tokens))
+        inputs = inputs + tuple(new_keys)
+        inputs_indices = inputs_indices + (None,) * len(new_keys)
+        kwargs = subs(kwargs, sub)
+        graph.update(dsk_kwargs)
+
+    indices = [(k, v) for k, v in zip(inputs, inputs_indices)]
+    keys = tuple(map(atop_token, range(len(inputs))))
+
+    if not kwargs:
+        dsk = {output: (func,) + keys}
+    else:
+        _keys = list(keys)
+        if new_keys:
+            _keys = _keys[:-len(new_keys)]
+        dsk = {output: (apply, func, _keys, kwargs)}
+
+    top = TOP(output, output_indices, dsk, indices,
+              numblocks=numblocks, concatenate=concatenate, new_axes=new_axes)
+    graph.update_with_key(top, output)
+    graph.dependencies = {output: {arg for arg, ind in argpairs if ind is not None}}
+    return graph
+
+
+class TOP(Mapping):
```

Comment:
Don't worry too much about this if dependencies are complicated, but if possible it might be nice to split the TOP/atop/fusion stuff all out into its own source file. `core.py` is getting rather long.

Reply:
Done.  I had to do a couple imports within functions, but overall I think that this is a good move